opal/mca/threads/pthreads/threads_pthreads_mutex.h

```diff
@@ -46,170 +47,163 @@
+        if (0 != ret) {
+            return OPAL_ERR_IN_ERRNO;
         }
+        pthread_mutexattr_settype(&mutex_attr, PTHREAD_MUTEX_RECURSIVE);
+        ret = pthread_mutex_init(p_mutex, &mutex_attr);
+        pthread_mutexattr_destroy(&mutex_attr);
+    } else {
+        ret = pthread_mutex_init(p_mutex, NULL);
+    }
 #endif
+    return 0 == ret ? OPAL_SUCCESS : OPAL_ERR_IN_ERRNO;
+}
 
-#if defined(OPAL_PTHREAD_RECURSIVE_MUTEX_INITIALIZER)
-
-#    if OPAL_ENABLE_DEBUG
-#        define OPAL_RECURSIVE_MUTEX_STATIC_INIT                                               \
-            {                                                                                  \
-                .super = OPAL_OBJ_STATIC_INIT(opal_mutex_t),                                   \
-                .m_lock_pthread = OPAL_PTHREAD_RECURSIVE_MUTEX_INITIALIZER, .m_lock_debug = 0, \
-                .m_lock_file = NULL, .m_lock_line = 0, .m_lock_atomic = OPAL_ATOMIC_LOCK_INIT, \
-            }
-#    else
-#        define OPAL_RECURSIVE_MUTEX_STATIC_INIT                            \
-            {                                                               \
-                .super = OPAL_OBJ_STATIC_INIT(opal_mutex_t),                \
-                .m_lock_pthread = OPAL_PTHREAD_RECURSIVE_MUTEX_INITIALIZER, \
-                .m_lock_atomic = OPAL_ATOMIC_LOCK_INIT,                     \
-            }
-#    endif
-
+static inline void opal_thread_internal_mutex_lock(opal_thread_internal_mutex_t *p_mutex)
+{
+#if OPAL_ENABLE_DEBUG
+    int ret = pthread_mutex_lock(p_mutex);
+    if (EDEADLK == ret) {
+        opal_output(0, "opal_thread_internal_mutex_lock() %d", ret);
+    }
+    assert(0 == ret);
+#else
+    pthread_mutex_lock(p_mutex);
 #endif
+}
 
-/************************************************************************
- *
- * mutex operations (non-atomic versions)
- *
- ************************************************************************/
-
-static inline int opal_mutex_trylock(opal_mutex_t *m)
+static inline int opal_thread_internal_mutex_trylock(opal_thread_internal_mutex_t *p_mutex)
 {
-    int ret = pthread_mutex_trylock(&m->m_lock_pthread);
-    if (EDEADLK == ret) {
+    int ret = pthread_mutex_trylock(p_mutex);
 #if OPAL_ENABLE_DEBUG
-        opal_output(0, "opal_mutex_trylock() %d", ret);
-#endif
+    if (EDEADLK == ret) {
+        opal_output(0, "opal_thread_internal_mutex_trylock() %d", ret);
         return 1;
     }
+#endif
     return 0 == ret ? 0 : 1;
+    ;
 }
 
-static inline void opal_mutex_lock(opal_mutex_t *m)
+static inline void opal_thread_internal_mutex_unlock(opal_thread_internal_mutex_t *p_mutex)
 {
 #if OPAL_ENABLE_DEBUG
-    int ret = pthread_mutex_lock(&m->m_lock_pthread);
+    int ret = pthread_mutex_unlock(p_mutex);
     if (EDEADLK == ret) {
-        errno = ret;
-        opal_output(0, "opal_mutex_lock() %d", ret);
+        opal_output(0, "opal_thread_internal_mutex_unlock() %d", ret);
```

Comment:
Not good, OMPI is not bound to Linux: https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_mutex_lock.html

Reply:
In my understanding, @devreal's point is that `_unlock()` and `_trylock()` do not return `EDEADLK` (the reference you mentioned is about `_lock()`:
```
The pthread_mutex_lock() function may fail if:
[EDEADLK]
  A deadlock condition was detected or the current thread already owns the mutex.
```

The POSIX specification does not mention the error code of those functions.

But other implementations (like FreeBSD or an unknown wild Linux-like OS) might return `EDEADLK`. Either is fine to me. Since keeping the current code is easy, I will keep it unless we all agree that we should remove this check.

Reply:
@bosilca From the link:

```
The pthread_mutex_lock() function may fail if:

[EDEADLK]
    A deadlock condition was detected or the current thread already owns the mutex. 
```
That is the only reference to `EDEADLK`. I don't see how `pthread_mutex_unlock` could return `EDEADLK`.

Reply:
you're right only _lock can return this error code.

Reply:
I removed this check for `unlock()`. Please check the last commit.